File: BikeHaus.Client/src/app/services/serviceleistung.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface Serviceleistung {
  id: number;
  belegNummer: string;
  datum: string;
  kundeName: string;
  kundeTelefon: string | null;
  kundeEmail: string | null;
  kundeAdresse: string | null;
  fahrradMarke: string | null;
  fahrradModell: string | null;
  rahmennummer: string | null;
  farbe: string | null;
  durchgefuehrteArbeiten: string;
  verwendeteTeile: string | null;
  preis: number | null;
  zahlungsart: string | null;
  notizen: string | null;
  createdAt?: string;
}

export interface ServiceleistungCreate {
  datum: string;
  kundeName: string;
  kundeTelefon: string | null;
  kundeEmail: string | null;
  kundeAdresse: string | null;
  fahrradMarke: string | null;
  fahrradModell: string | null;
  rahmennummer: string | null;
  farbe: string | null;
  durchgefuehrteArbeiten: string;
  verwendeteTeile: string | null;
  preis: number | null;
  zahlungsart: string | null;
  notizen: string | null;
}

@Injectable({
  providedIn: 'root',
})
export class ServiceleistungService {
  private apiUrl = `${environment.apiUrl}/serviceleistungen`;

  constructor(private http: HttpClient) {}

  getAll(): Observable<Serviceleistung[]> {
    return this.http.get<Serviceleistung[]>(this.apiUrl);
  }

  getById(id: number): Observable<Serviceleistung> {
    return this.http.get<Serviceleistung>(`${this.apiUrl}/${id}`);
  }

  create(item: ServiceleistungCreate): Observable<Serviceleistung> {
    return this.http.post<Serviceleistung>(this.apiUrl, item);
  }

  update(id: number, item: ServiceleistungCreate): Observable<Serviceleistung> {
    return this.http.put<Serviceleistung>(`${this.apiUrl}/${id}`, item);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getNextBelegNummer(): Observable<{ belegNummer: string }> {
    return this.http.get<{ belegNummer: string }>(
      `${this.apiUrl}/next-belegnummer`,
    );
  }

  downloadServicebeleg(id: number): Observable<Blob> {
    return this.http.get(`${this.apiUrl}/${id}/servicebeleg`, {
      responseType: 'blob',
    });
  }
}

File: BikeHaus.Application/Services/ServiceleistungService.cs
using BikeHaus.Application.DTOs;
using BikeHaus.Application.Interfaces;
using BikeHaus.Domain.Entities;
using BikeHaus.Domain.Interfaces;

namespace BikeHaus.Application.Services;

public class ServiceleistungService : IServiceleistungService
{
    private readonly IRepository<Serviceleistung> _repository;

    public ServiceleistungService(IRepository<Serviceleistung> repository)
    {
        _repository = repository;
    }

    public async Task<IEnumerable<ServiceleistungDto>> GetAllAsync()
    {
        var items = await _repository.GetAllAsync();
        return items
            .OrderByDescending(s => s.Datum)
            .ThenByDescending(s => s.Id)
            .Select(MapToDto);
    }

    public async Task<ServiceleistungDto?> GetByIdAsync(int id)
    {
        var item = await _repository.GetByIdAsync(id);
        return item == null ? null : MapToDto(item);
    }

    public async Task<ServiceleistungDto> CreateAsync(ServiceleistungCreateDto dto)
    {
        var item = new Serviceleistung
        {
            BelegNummer = await GetNextBelegNummerAsync(),
            Datum = dto.Datum,
            KundeName = dto.KundeName,
            KundeTelefon = dto.KundeTelefon,
            KundeEmail = dto.KundeEmail,
            KundeAdresse = dto.KundeAdresse,
            FahrradMarke = dto.FahrradMarke,
            FahrradModell = dto.FahrradModell,
            Rahmennummer = dto.Rahmennummer,
            Farbe = dto.Farbe,
            DurchgefuehrteArbeiten = dto.DurchgefuehrteArbeiten,
            VerwendeteTeile = dto.VerwendeteTeile,
            Preis = dto.Preis,
            Zahlungsart = dto.Zahlungsart,
            Notizen = dto.Notizen
        };

        var created = await _repository.AddAsync(item);
        return MapToDto(created);
    }

    public async Task<ServiceleistungDto?> UpdateAsync(int id, ServiceleistungUpdateDto dto)
    {
        var item = await _repository.GetByIdAsync(id);
        if (item == null) return null;

        item.Datum = dto.Datum;
        item.KundeName = dto.KundeName;
        item.KundeTelefon = dto.KundeTelefon;
        item.KundeEmail = dto.KundeEmail;
        item.KundeAdresse = dto.KundeAdresse;
        item.FahrradMarke = dto.FahrradMarke;
        item.FahrradModell = dto.FahrradModell;
        item.Rahmennummer = dto.Rahmennummer;
        item.Farbe = dto.Farbe;
        item.DurchgefuehrteArbeiten = dto.DurchgefuehrteArbeiten;
        item.VerwendeteTeile = dto.VerwendeteTeile;
        item.Preis = dto.Preis;
        item.Zahlungsart = dto.Zahlungsart;
        item.Notizen = dto.Notizen;
        item.UpdatedAt = DateTime.UtcNow;

        await _repository.UpdateAsync(item);
        return MapToDto(item);
    }

    public async Task<bool> DeleteAsync(int id)
    {
        var item = await _repository.GetByIdAsync(id);
        if (item == null) return false;

        await _repository.DeleteAsync(id);
        return true;
    }

    // Format: SL-2026-0001, fortlaufend pro Jahr
    public async Task<string> GetNextBelegNummerAsync()
    {
        var year = DateTime.UtcNow.Year;
        var prefix = $"SL-{year}-";
        var existing = await _repository.FindAsync(s => s.BelegNummer.StartsWith(prefix));

        var maxSeq = 0;
        foreach (var s in existing)
        {
            var tail = s.BelegNummer.Substring(prefix.Length);
            if (int.TryParse(tail, out var seq) && seq > maxSeq)
                maxSeq = seq;
        }

        return $"{prefix}{maxSeq + 1:D4}";
    }

    private static ServiceleistungDto MapToDto(Serviceleistung s) => new(
        s.Id, s.BelegNummer, s.Datum,
        s.KundeName, s.KundeTelefon, s.KundeEmail, s.KundeAdresse,
        s.FahrradMarke, s.FahrradModell, s.Rahmennummer, s.Farbe,
        s.DurchgefuehrteArbeiten, s.VerwendeteTeile,
        s.Preis, s.Zahlungsart, s.Notizen, s.CreatedAt
    );
}
